File: Abstract Factory/Exercicios/Ex4_Robos/Ex4.js
/**
 * PADRÃO: ABSTRACT FACTORY (FÁBRICA ABSTRATA)
 * 
 * Exercício: Sistema de Fábrica de Robôs.
 * Este exercício mostra como as fábricas abstratas podem garantir que as partes 
 * de um robô (Armadura e Arma) sejam compatíveis com o ambiente de destino.
 */

// 1. PRODUTOS ABSTRATOS
class Armadura {
    exibir() { throw new Error("Método abstrato"); }
}

class Arma {
    exibir() { throw new Error("Método abstrato"); }
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA TERRESTRE
class ArmaduraTanque extends Armadura {
    exibir() { return "Armadura: Blindagem Pesada (Terrestre)."; }
}

class Canhao extends Arma {
    exibir() { return "Arma: Canhão de Longo Alcance."; }
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA ESPACIAL
class ArmaduraLeve extends Armadura {
    exibir() { return "Armadura: Revestimento de Titânio (Espacial)."; }
}

class Laser extends Arma {
    exibir() { return "Arma: Canhão Laser de Precisão."; }
}

// 3. FABRICA ABSTRATA
class FabricaRobo {
    criarArmadura() { throw new Error("Método abstrato"); }
    criarArma() { throw new Error("Método abstrato"); }
}

// 4. FABRICAS CONCRETAS
class FabricaRoboTerrestre extends FabricaRobo {
    criarArmadura() { return new ArmaduraTanque(); }
    criarArma() { return new Canhao(); }
}

class FabricaRoboEspacial extends FabricaRobo {
    criarArmadura() { return new ArmaduraLeve(); }
    criarArma() { return new Laser(); }
}

// 5. CLIENTE
function montarRobo(fabrica) {
    const armadura = fabrica.criarArmadura();
    const arma = fabrica.criarArma();

    console.log(armadura.exibir());
    console.log(arma.exibir());
    console.log(" ----------------------------");
}

console.log("\n--- CONFIGURAÇÃO TERRESTRE ---");
montarRobo(new FabricaRoboTerrestre());

console.log("\n--- CONFIGURAÇÃO ESPACIAL ---");
montarRobo(new FabricaRoboEspacial());

File: Abstract Factory/Exercicios/Ex4_Robos/Ex4.java
/**
 * PADRÃO: ABSTRACT FACTORY (FÁBRICA ABSTRATA)
 * 
 * Exercício: Sistema de Fábrica de Robôs.
 * O objetivo é criar famílias de componentes (Armadura e Arma) 
 * que variam de acordo com o ambiente de operação (Terrestre ou Espacial).
 */

// 1. PRODUTOS ABSTRATOS
interface Armadura {
    String exibir();
}

interface Arma {
    String exibir();
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA TERRESTRE
class ArmaduraTanque implements Armadura {
    public String exibir() { return "Armadura: Blindagem Pesada (Terrestre)."; }
}

class Canhao implements Arma {
    public String exibir() { return "Arma: Canhão de Longo Alcance."; }
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA ESPACIAL
class ArmaduraLeve implements Armadura {
    public String exibir() { return "Armadura: Revestimento de Titânio (Espacial)."; }
}

class Laser implements Arma {
    public String exibir() { return "Arma: Canhão Laser de Precisão."; }
}

// 3. FABRICA ABSTRATA
interface FabricaRobo {
    Armadura criarArmadura();
    Arma criarArma();
}

// 4. FABRICAS CONCRETAS
class FabricaRoboTerrestre implements FabricaRobo {
    public Armadura criarArmadura() { return new ArmaduraTanque(); }
    public Arma criarArma() { return new Canhao(); }
}

class FabricaRoboEspacial implements FabricaRobo {
    public Armadura criarArmadura() { return new ArmaduraLeve(); }
    public Arma criarArma() { return new Laser(); }
}

// 5. CLIENTE
public class Ex4 {
    public static void montarRobo(FabricaRobo fabrica) {
        Armadura armadura = fabrica.criarArmadura();
        Arma arma = fabrica.criarArma();

        System.out.println(armadura.exibir());
        System.out.println(arma.exibir());
        System.out.println(" ----------------------------");
    }

    public static void main(String[] args) {
        System.out.println("\n--- CONFIGURAÇÃO TERRESTRE ---");
        montarRobo(new FabricaRoboTerrestre());

        System.out.println("\n--- CONFIGURAÇÃO ESPACIAL ---");
        montarRobo(new FabricaRoboEspacial());
    }
}
